### PR TITLE
[VarExporter] fix contravariance problem with __unserialize() in lazy proxy

### PR DESCRIPTION
**Pull request by nikophil**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Fix | #57470
| License       | MIT

`LazyProxyTrait::__unserialize()` has currently the following prototype: `public function __unserialize(array $data): void`. So, classes defining this `__unserialize()` method without the `array` type hint could not use `LazyProxyTrait`

```
Compile Error: Declaration of Symfony\Component\VarExporter\LazyProxyTrait::__unserialize(array $data): void must be compatible with Foo::__unserialize($data)
```

### CODE REVIEW

**Comment by derrabus** [comment]
Not adding types will cause trouble in other scenarios. We should probably document how to use the trait in untyped codebases.

**Comment by nicolas-grekas** [comment]
Instead, we could generate the method in ProxyHelper maybe?
Something like
```php
use LazyProxyTrait {
    __unserialize as __doUnserialize;
}

public function __unserialize($data): void { $this->__doUnserialize($data); }
```

**Comment by stof** [comment]
wouldn't this cause issues in those other cases mentioned above ? And if no, why not updating the trait directly ?

File: src/Symfony/Component/VarExporter/ProxyHelper.php [hunk 1]
```diff
@@ -200,10 +200,17 @@ public static function generateLazyProxy(?\ReflectionClass $class, array $interf
         return <<<EOPHP
             {$parent} implements \\{$interfaces}
             {
-                use \Symfony\Component\VarExporter\LazyProxyTrait;
+                use \Symfony\Component\VarExporter\LazyProxyTrait {
+                    __unserialize as __doUnserialize;
```
Comment (stof) [inline src/Symfony/Component/VarExporter/ProxyHelper.php]: `__doUnserialize` should probably be `private`

**Comment by nikophil** [comment]
> Not adding types will cause trouble in other scenarios

as a curiosity, which would be these scenarios?

> Instead, we could generate the method in ProxyHelper maybe?

done!

**Comment by nicolas-grekas** [comment]
> why not updating the trait directly ?

for correctness and also to allow child classes to add the type hint I'd say

File: src/Symfony/Component/VarExporter/ProxyHelper.php [hunk 2]
```diff
@@ -200,10 +200,17 @@ public static function generateLazyProxy(?\ReflectionClass $class, array $interf
...
+                }
 
                 private const LAZY_OBJECT_PROPERTY_SCOPES = {$propertyScopes};
-            {$body}}
+            {$body}
+                public function __unserialize(\$data): void
+                { 
+                    \$this->__doUnserialize(\$data);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarExporter/ProxyHelper.php]: I'd do all this only if the parent class doesn't have a type-hint!

Comment (nikophil) [reply src/Symfony/Component/VarExporter/ProxyHelper.php]: done

**Comment by nicolas-grekas** [comment]
Tests are failing so there are some details to fix but this looks good to me so far, thanks.

**Comment by nikophil** [comment]
yes, I saw this... any hints about the failure?
It occurs in doctrine bridge tests: variable `$service` [here](https://github.com/doctrine/persistence/blob/3.3.x/src/Persistence/AbstractManagerRegistry.php#L165) seems to be a `stdClass_c562334` (hence the failure). But I have the same behavior on branch`6.4` :thinking:

**Comment by derrabus** [comment]
The PR title does not reflect the proposed change anymore though.

File: src/Symfony/Component/VarExporter/ProxyHelper.php [hunk 3]
```diff
@@ -197,10 +197,36 @@ public static function generateLazyProxy(?\ReflectionClass $class, array $interf
         $body = $methods ? "\n".implode("\n\n", $methods)."\n" : '';
         $propertyScopes = $class ? self::exportPropertyScopes($class->name) : '[]';
 
+        if (
+            $class?->hasMethod('__unserialize')
+            && null !== ($unserializeParameter = $class->getMethod('__unserialize')->getParameters()[0] ?? null)
+            && null === $unserializeParameter->getType()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarExporter/ProxyHelper.php]: you spotted it already: the engine checks that there is one parameter so we don't need the `?? null` part ;)
```suggestion
            && $class->getMethod('__unserialize')->getParameters()[0]->getType()
```

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/VarExporter/ProxyHelper.php [hunk 4]
```diff
@@ -197,10 +197,35 @@ public static function generateLazyProxy(?\ReflectionClass $class, array $interf
         $body = $methods ? "\n".implode("\n\n", $methods)."\n" : '';
         $propertyScopes = $class ? self::exportPropertyScopes($class->name) : '[]';
 
+        if (
+            $class?->hasMethod('__unserialize')
+            && null === $class->getMethod('__unserialize')->getParameters()[0]->getType()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarExporter/ProxyHelper.php]:
```suggestion
            && !$class->getMethod('__unserialize')->getParameters()[0]->getType()
```

**Comment by nicolas-grekas** [comment]
Thank you @nikophil.

**Pull request merged**